File: starters/python/tools/gen_manifest.py
#!/usr/bin/env python3
"""Generate manifest.json from the REAL module graph — never hand-authored.

manifest.json records:
  - capabilities: every public top-level function/class -> the one file
    that defines it (derived from core/, domains/*, app/ source, excluding
    tests).
  - edges: every project-local "module A imports module B" edge, derived
    the same way tools/boundary_check.py derives its view of the graph.

Run with --check to verify the committed manifest.json still matches what
the source tree would generate (the CI drift-check gate: fails if code
changed but nobody ran `make manifest`).
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from tools._pygraph import (
    discover_py_files,
    module_name_for_path,
    parse_imports_raw,
    repo_root,
    resolve_project_local,
    top_level_defs,
)

MANIFEST_PATH_NAME = "manifest.json"


def build_manifest(root: Path) -> dict:
    files = discover_py_files(root, include_tests=False)
    known_modules = {module_name_for_path(root, p) for p in files}

    capabilities = {}
    for path in files:
        rel = path.relative_to(root).as_posix()
        module = module_name_for_path(root, path)
        for name, _lineno, kind in top_level_defs(path):
            if name.startswith("_"):
                continue
            capabilities[f"{module}.{name}"] = {"file": rel, "kind": kind}

    edges = set()
    for path in files:
        module = module_name_for_path(root, path)
        seen = set()
        for candidate in parse_imports_raw(path):
            target = resolve_project_local(candidate, known_modules)
            if target is None or target == module or target in seen:
                continue
            seen.add(target)
            edges.add((module, target))

    return {
        "generated_by": "tools/gen_manifest.py (auto-generated - do not hand-edit)",
        "capabilities": dict(sorted(capabilities.items())),
        "edges": [
            {"from": a, "to": b} for a, b in sorted(edges)
        ],
    }


def render(manifest: dict) -> str:
    return json.dumps(manifest, indent=2, sort_keys=True) + "\n"


def main() -> int:
    root = repo_root()
    manifest = build_manifest(root)
    rendered = render(manifest)
    manifest_path = root / MANIFEST_PATH_NAME

    if "--check" in sys.argv[1:]:
        if not manifest_path.exists():
            print(f"manifest_check: {MANIFEST_PATH_NAME} does not exist. "
                  f"Run `make manifest` and commit it.")
            return 1
        current = manifest_path.read_text(encoding="utf-8")
        if current != rendered:
            print(f"manifest_check: {MANIFEST_PATH_NAME} is stale — the "
                  f"module graph changed but the manifest was not "
                  f"regenerated. Run `make manifest` and commit the diff.")
            import difflib
            diff = difflib.unified_diff(
                current.splitlines(keepends=True),
                rendered.splitlines(keepends=True),
                fromfile=f"{MANIFEST_PATH_NAME} (committed)",
                tofile=f"{MANIFEST_PATH_NAME} (regenerated)",
            )
            sys.stdout.writelines(diff)
            return 1
        print(f"manifest_check: {MANIFEST_PATH_NAME} is up to date "
              f"({len(manifest['capabilities'])} capabilities, "
              f"{len(manifest['edges'])} edges)")
        return 0

    manifest_path.write_text(rendered, encoding="utf-8")
    print(f"wrote {MANIFEST_PATH_NAME} "
          f"({len(manifest['capabilities'])} capabilities, "
          f"{len(manifest['edges'])} edges)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: starters/typescript/tools/gen-manifest.mjs
#!/usr/bin/env node
// Generates manifest.json — a capability-to-owning-file map plus the
// package/file dependency edges — directly from the real module graph
// under packages/. Nothing here is hand-authored: run with --write to
// regenerate it, or --check to verify the committed file still matches
// the graph (used as CI's drift check).
//
// Output is fully deterministic (sorted, no timestamps) so that
// `--write` twice in a row with no code changes produces a byte-identical
// file, and `--check` only ever fails when the graph actually changed.
import { writeFileSync, readFileSync, existsSync } from "node:fs";
import path from "node:path";
import {
  listPackages,
  listSourceFiles,
  packageOf,
  parseImportSpecifiers,
  parseExportedDeclarations,
  resolveRelativeImport,
  readFile,
  toRepoRelative,
  REPO_ROOT,
  PACKAGES_ROOT,
} from "./lib/scan.mjs";

const MANIFEST_PATH = path.join(REPO_ROOT, "manifest.json");

function buildManifest() {
  const packages = listPackages();
  const files = listSourceFiles();

  // capabilities: every exported top-level declaration under packages/*/src
  // (test files export test setup, not product capabilities, so they're
  // excluded here but still counted in the edges below).
  const capabilities = [];
  for (const file of files) {
    const pkg = packageOf(file);
    const isSrc = toRepoRelative(file).split("/").includes("src");
    if (!isSrc) continue;
    for (const decl of parseExportedDeclarations(readFile(file))) {
      capabilities.push({
        capability: decl.name,
        kind: decl.kind,
        package: pkg,
        file: toRepoRelative(file),
      });
    }
  }
  capabilities.sort((a, b) => (a.capability + a.file).localeCompare(b.capability + b.file));

  // fileEdges / packageEdges: every relative import between two files
  // under packages/, resolved from the real source text.
  const fileEdgeSet = new Set();
  const fileEdges = [];
  const packageEdgeSet = new Set();
  const packageEdges = [];

  for (const file of files) {
    const fromPkg = packageOf(file);
    const source = readFile(file);
    for (const specifier of parseImportSpecifiers(source)) {
      const resolved = resolveRelativeImport(file, specifier);
      if (resolved === null) continue;
      const toPkg = packageOf(resolved);
      if (toPkg === null) continue;

      const fileKey = `${toRepoRelative(file)} -> ${toRepoRelative(resolved)}`;
      if (!fileEdgeSet.has(fileKey)) {
        fileEdgeSet.add(fileKey);
        fileEdges.push({ from: toRepoRelative(file), to: toRepoRelative(resolved) });
      }

      if (fromPkg !== toPkg) {
        const pkgKey = `${fromPkg} -> ${toPkg}`;
        if (!packageEdgeSet.has(pkgKey)) {
          packageEdgeSet.add(pkgKey);
          packageEdges.push({ from: fromPkg, to: toPkg });
        }
      }
    }
  }

  fileEdges.sort((a, b) => (a.from + a.to).localeCompare(b.from + b.to));
  packageEdges.sort((a, b) => (a.from + a.to).localeCompare(b.from + b.to));

  const manifest = {
    generatedBy: "tools/gen-manifest.mjs",
    note: "Generated from the packages/ module graph. Do not hand-edit — run `npm run generate:manifest`.",
    packages,
    capabilities,
    packageEdges,
    fileEdges,
  };

  return JSON.stringify(manifest, null, 2) + "\n";
}

function main() {
  const mode = process.argv[2];
  const generated = buildManifest();

  if (mode === "--write") {
    writeFileSync(MANIFEST_PATH, generated, "utf8");
    console.log(`gen-manifest: wrote ${toRepoRelative(MANIFEST_PATH)}`);
    return;
  }

  if (mode === "--check") {
    if (!existsSync(MANIFEST_PATH)) {
      console.error(
        `gen-manifest: ${toRepoRelative(MANIFEST_PATH)} does not exist. Run \`npm run generate:manifest\`.`,
      );
      process.exit(1);
    }
    const committed = readFileSync(MANIFEST_PATH, "utf8");
    if (committed !== generated) {
      console.error(
        "gen-manifest: manifest.json is out of date with packages/ — code changed but the manifest was not regenerated.\n" +
          "Run `npm run generate:manifest` and commit the result.",
      );
      process.exit(1);
    }
    console.log("gen-manifest: manifest.json matches the current module graph (no drift).");
    return;
  }

  console.error("usage: node tools/gen-manifest.mjs --write | --check");
  process.exit(2);
}

main();
